modules/library/metadata/src/main/java/org/geotools/factory/CategoryRegistry.java

```diff
@@ -0,0 +1,301 @@
+package org.geotools.factory;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMap.Builder;
+import org.geotools.util.PartiallyOrderedSet;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+import static org.geotools.util.Utilities.stream;
+
+/**
+ * The category registry holds multiple instances per category. Categories are
+ * {@link Class classes} and instances are also accessible by the class they implement.
+ * Note that instances have to implement/extend the category they are filed under.
+ */
+class CategoryRegistry {
+
+	/**
+	 * Heterogeneous map, where `Class<T>` is mapped to `InstanceRegistry<T>` for any `T`.
+	 *
+	 * Code depends on this map being immutable to avoid concurrent modification exceptions.
+	 */
+	private final ImmutableMap<Class<?>, InstanceRegistry<?>> categories;
+
+	/**
+	 * Creates a new registry with the specified Registers the specified category.
+	 * If the same category is registered multiple times, all instances previously
+	 * registered for that category are lost.
+	 *
+	 * @param factoryRegistry The {@link FactoryRegistry} this registry belongs to.
+	 * @param categories The categories to register; must not be {@code null} but can contain {@code null}.
+	 */
+	public CategoryRegistry(final FactoryRegistry factoryRegistry, final Iterable<Class<?>> categories) {
+		requireNonNull(factoryRegistry);
+		requireNonNull(categories);
+		this.categories = createCategoriesMap(factoryRegistry, categories);
+	}
+
+	private static ImmutableMap<Class<?>, InstanceRegistry<?>> createCategoriesMap(
+			final FactoryRegistry factoryRegistry, final Iterable<Class<?>> categories) {
+		Builder<Class<?>, InstanceRegistry<?>> categoriesBuilder = ImmutableMap.builder();
+		categories.forEach(category ->
+				categoriesBuilder.put(category, new InstanceRegistry<>(factoryRegistry, category)));
+		return categoriesBuilder.build();
+	}
+
+	/**
+	 * Registers the specified instance under all categories that are supertypes of the instance.
+	 *
+	 * @param instance The instance to register.
+	 */
+	@SuppressWarnings("unchecked")
+	public <T> void registerInstance(final T instance) {
+		requireNonNull(instance);
+		streamCategories()
+				.filter(category -> category.isAssignableFrom(instance.getClass()))
+				// the cast is correct because the filter above only leaves categories that are supertypes of `instance`
+				.map(category -> (InstanceRegistry<? super T>) instanceRegistry(category))
```

Comment:
You may consider passing in Class<T> type as a parameter to avoid this cast

Reply:
Do you mean have a second parameter `Class<T>`? I don't see how that would spare me the cast.

Also, I don't see how we could pass that parameter. The calling method `FactoryRegistry::registerFactory` only has the instance as an `Object`. Even if we generified that method...

    public <T> void registerFactory(final T factory)

... it wouldn't work because `factory.getClass()` returns `Class<? extends Object>`.